### ISSUE TITLE
atomic function load() on GPU

### ISSUE THREAD

**Issue by arkantos493** [illuhad/hipSYCL#255]
Calling `load()` on an atomic sycl accessor targeting the GPU (NVIDIA) result in a linker error:
```
$ syclcc-clang -O3 --hipsycl-platform=cuda main.cpp
ptxas fatal   : Unresolved extern function '_ZNVK7hipsycl4sycl6atomicIiLNS0_6access13address_spaceE0EE4loadENS0_12memory_orderE'
clang-10: error: ptxas command failed with exit code 255 (use -v to see invocation)
```

```cpp
#include <vector>

#include <CL/sycl.hpp>

namespace sycl = cl::sycl;

int main() {
    std::vector<int> vec(10, 0);

    {
        sycl::queue queue(sycl::default_selector{});
        std::cout << "Used device: " << queue.get_device().get_info<sycl::info::device::name>() << '\n' << std::endl;
        
        sycl::buffer<int, 1> buffer(vec.data(), vec.size());

        queue.submit([&](sycl::handler& cgh) {
            auto acc = buffer.get_access<sycl::access::mode::atomic>(cgh);
            
            cgh.parallel_for<class test_kernel>(sycl::range<>(vec.size()), [=](sycl::item<> item) {
                const auto idx = item.get_linear_id();
                acc[idx].load();  // <-
            });
        });
    } 

    return 0;
}
```
Used syclcc-clang version:
```
$ syclcc-clang --version
clang version 10.0.0 
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: Programs/spack/opt/spack/linux-linuxmint19-zen/gcc-8.4.0/llvm-10.0.0-uxnrorpisttivegjr4eqshrdva3heg55/bin 
```

**Comment by fodinabor** [comment]
That's probably known as atomics on GPU are not implemented in master, see e.g. issue #213

**Comment by illuhad** [comment]
Yes, on GPU `load()` and `store()` are not implemented because it's unclear how to do it efficiently in HIP. If you know how to do it, please submit a pull request :)

**Comment by arkantos493** [comment]
Sorry I completely forgot that issue.

**Comment by FilipVaverka** [comment]
Could we simply use [GNU atomic builtins](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) to solve this problem?
I've checked disassembly of "__atomic_load_n" and "__atomic_store_n" used in "__global__" region in code compiled for gfx900 with hipcc and it seems the assembly matches [memory model requirements](https://llvm.org/docs/AMDGPUUsage.html#memory-model-gfx6-gfx9). While HIP doesn't expose these directly it does use them internally to implement its own [atomic functions](https://github.com/ROCm-Developer-Tools/HIP/blob/main/include/hip/hcc_detail/hip_atomic.h).

However, I don't know if these are implemented for CUDA.

**Comment by illuhad** [comment]
@FilipVaverka That's good to know. We already use them on CPU, so we could also enable them as well in the HIP path. 
Last time we checked there was no such functionality in CUDA, but it seems libcudacxx does seem to have something like it:
https://github.com/NVIDIA/libcudacxx/blob/7ef6365926ad1e13ed8ff8b320bdc359457923e0/include/cuda/std/detail/__atomic#L224

However I haven't found out yet where `__atomic_load_n_cuda` comes from and whether this would also work with clang-compiled CUDA.
